External file uploader

AckslD opened the pull request:
this is a rebase of https://github.com/qutebrowser/qutebrowser/pull/4477, resolving conflicts

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +342,39 @@ def netrc_authentication(url, authenticator):
+    proc.finished.connect(lambda _code, _status: loop.exit())
+    loop.exec_()
+
+    with open(filename, 'r', encoding='utf8') as tmpfile:
```
Not fully sure what mypy is complaining about here

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +342,39 @@ def netrc_authentication(url, authenticator):
+    if not multiple:
+        assert len(selected_file_list) <= 1
+
+    return selected_file_list
```
Not sure why mypy thinks this is a `List[bytes]`. If I print `type(selected_file_list[0])` I get `str`.

AckslD commented:
@The-Compiler or someone else know how to solve the mypy issues above?

The-Compiler commented:
The main reason is that you're using `tmpfile` for two different things:

- `with tempfile.NamedTemporaryFile(delete=False) as tmpfile:` (which uses `mode='w+b` by default, so gives you a binary rather than a text file)
- `with open(filename, 'r', encoding='utf8') as tmpfile:` (which is now a text file rather than a binary one)

And mypy doesn't support using the same name for two different types (which is probably a good thing).

However, I suspect there's a much cleaner way to do the tempfile handling, I don't really see the point of using `delete=False` and then doing the deletion manually. I'll follow-up on this with a bit more detail in a couple of days, tuning out of work and qutebrowser for a bit at the moment :slightly_smiling_face:

AckslD commented:
Ah, I see! Thanks I think I understand now. I'll try to change and clean this up. Enjoy some time of :)

AckslD commented:
@The-Compiler For when you're back, I changed how the temporary file is handled, let me know what you think. I also changed so the encoding used is taken from `sys.getfilesystemencoding`.

There are some tests failing randomly due to some timeout, does that happen also on master?

The-Compiler commented:
The `test_resource_with_invalid_url` failure also happened on master - not sure what causes it to show up so much suddenly, but I tried to stabilize things in bde129b4cfea28d47e9f1e2e8b36f7ecd628c04b. If you want to rebase on the latest master before I review this, that'd be great.

As for the macOS 11.0 issue, that seems like an issue on GitHub's side. Feel free to ignore it.

AckslD commented:
Cool, I've rebased on master now

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -238,3 +238,12 @@ def acceptNavigationRequest(self,
+    def chooseFiles(self, mode, old_files, accepted_mimetypes):
+        """Override chooseFiles to (optionally) invoke custom file uploader."""
+        handler = config.val.fileselect.handler
+        if handler == "default":
```
Nitpick: Simpler to use the config value directly here I'd say:

```suggestion
        if config.val.fileselect.handler == "default":
```

AckslD replied:
Fixed manually

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -238,3 +238,12 @@ def acceptNavigationRequest(self,
         self.navigation_request.emit(navigation)
         return navigation.accepted
+
+    def chooseFiles(self, mode, old_files, accepted_mimetypes):
```
Mind adding type annotations here as well? Something like:

```suggestion
    def chooseFiles(self, mode: QWebEnginePage.FileSelectionMode, old_files: List[str], accepted_mimetypes: List[str]) -> List[str]:
```

(untested)

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -192,6 +192,18 @@ def _handle_errorpage(self, info, errpage):
             return True
 
+    def chooseFile(self, parent_frame: QWebFrame, suggested_file: str) -> str:
+        """Override chooseFiles to (optionally) invoke custom file uploader."""
```
Nitpick:

```suggestion
        """Override chooseFile to (optionally) invoke custom file uploader."""
```

(or drop "Override chooseFile to" entirely, both for QtWebEngine above and here)

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -192,6 +192,18 @@ def _handle_errorpage(self, info, errpage):
+    def chooseFile(self, parent_frame: QWebFrame, suggested_file: str) -> str:
+        """Override chooseFiles to (optionally) invoke custom file uploader."""
+        handler = config.val.fileselect.handler
+        if handler == "default":
```
```suggestion
        if config.val.fileselect.handler == "default":
```

AckslD replied:
Fixed manually

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -205,13 +217,17 @@ def _handle_multiple_files(self, info, files):
-        files.fileNames, _ = QFileDialog.getOpenFileNames(
-            None, None, suggested_file)  # type: ignore[arg-type]
+        handler = config.val.fileselect.handler
+        if handler == "default":
```
```suggestion
        if config.val.fileselect.handler == "default":
```

AckslD replied:
Fixed manually

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -1173,6 +1173,44 @@ editor.encoding:
+    placeholder: true
+  default: ['xterm', '-e', 'export COLUMNS= LINES=;ranger --choosefile={}']
+  desc: >-
+    Command (and arguments) to use for selecting a file in forms.
```
```suggestion
    Command (and arguments) to use for selecting a single file in forms.
```

To make it a bit clearer maybe?

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -1173,6 +1173,44 @@ editor.encoding:
+    valid_values:
+      - default: "Use the default file selector."
+      - external: "Use an external command."
+  desc: Handler for selecting file(s) in forms.
```
Maybe add something pointing to `fileselect.single_file.command` and `fileselect.multiple_files.command` here - see https://github.com/qutebrowser/qutebrowser/pull/4477#discussion_r251257562 (I'd probably add it in the setting description here, rather than in the value description above).

AckslD replied:
Updated description to point to `fileselect.single_file.command` and `fileselect.multiple_files.command`

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -1173,6 +1173,44 @@ editor.encoding:
+  type:
+    name: ShellCommand
+    placeholder: true
+  default: ['xterm', '-e', 'export COLUMNS= LINES=;ranger --choosefile={}']
```
I'm guessing you got the `export COLUMNS= LINES=` from https://github.com/qutebrowser/qutebrowser/pull/4477#issuecomment-451069864 - right? I've just tried this with `ranger` and it displays as 80x25 despite this. Not sure why that's the case, but given that the `export COLUMNS= LINES=;` doesn't actually seem to help, I suppose we might as well just drop it in the default value here?

AckslD replied:
I haven't actually tried it with ranger as of yet, but with `vifm`. Do you have a preference with `ranger` or could we also have the default use `vifm`? For example as:
```
['xterm', '-e', 'vifm', '--choose-file', '{}']
```

The-Compiler replied:
No strong preferences from my side here. I'm guessing ranger is much more popular:

- GitHub stars: [ranger](https://github.com/ranger/ranger/) 8.9k, [vifm](https://github.com/vifm/vifm) 1.4k
- Archlinux package stats: [ranger](https://pkgstats.archlinux.de/packages/ranger) ~20%, [vifm](https://pkgstats.archlinux.de/packages/vifm) ~8.5%
- Debian package stats: [ranger](https://qa.debian.org/popcon.php?package=ranger) 1.02%, [vifm](https://qa.debian.org/popcon.php?package=vifm) 0.25%

So some 2-6x times more common - at the same time, I suppose the default here isn't really a suitable default for most people anyways, and many will probably end up changing it (to e.g. use another terminal rather than `xterm`).

I'd kinda prefer keeping ranger because it's probably the more suitable starting point for most people, but if you want to go for vifm instead, I'm not going to stop you :slightly_smiling_face: 

Maybe it'd be a good idea to add a `completions:` to be able to complete ranger/vifm in xterm though?

AckslD replied:
Sounds good :+1:! I added completions to the commands. Let me know if that's the way to add completions and the way you meant :)

AckslD replied:
Hmm, it gives an error. I was looking at `hints.chars` for how to add `completions:` but maybe there is something missing?

The-Compiler reviewed (changes requested):
Thank you! Glad to see someone picking this up - I've left a couple of comments.

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,38 @@ def netrc_authentication(url, authenticator):
+    """
+    encoding = sys.getfilesystemencoding()
+    with tempfile.NamedTemporaryFile(mode='w+', encoding=encoding) as tmpfile:
+        filename = tmpfile.name
```
Changed

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,38 @@ def netrc_authentication(url, authenticator):
+                   [arg.replace('{}', filename) for arg in command[1:]])
+
+        loop = qtutils.EventLoop()
+        proc.finished.connect(lambda _code, _status: loop.exit())
```
I've been trying it with vifm and simply closing it without selecting files provides an empty list. I haven't tried with ranger. What would be a good way to catch unsuccessful exits?

The-Compiler replied:
I'm not sure actually how we'd handle an unsuccessful exit properly. With QtWebKit we can say "selecting files wasn't successful", but we can't do that with QtWebEngine.

Looking at the QtWebEngine code, it'll call `rejected()` on the `FilePickerController` (which I assume means "cancelled") if the list of selected filenames is empty, so I suppose the current handling is sufficient. The only case where it'd make a difference is if the spawned file manager would crash somehow after having written a filename (or filenames). I suppose that's rather uncommon, so let's leave it as-is.

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,38 @@ def netrc_authentication(url, authenticator):
+        selected_file_list = tmpfile.read().splitlines()
+
+    if not multiple:
+        assert len(selected_file_list) <= 1
```
Fixed

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,38 @@ def netrc_authentication(url, authenticator):
+        proc.finished.connect(lambda _code, _status: loop.exit())
+        loop.exec_()
+
+        selected_file_list = tmpfile.read().splitlines()
```
Nitpick fixed.

For windows let me know if it works, I don't have access to windows at the moment.

AckslD replied:
I now see the comment regarding window you linked to. Should we revert to the old/non-clean way or would you have some other idea how to handle it?

AckslD commented:
Thanks for the comments @The-Compiler! I fixed them except two outstanding which I left some questions.

The-Compiler commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,37 @@ def netrc_authentication(url, authenticator):
+
+    if not multiple:
+        if len(selected_files) > 1:
+            log.misc.exception("More than one files choosen, using only the first")
```
```suggestion
            message.warning("More than one files choosen, using only the first")
```

So that the warning actually gets displayed in the user interface.

AckslD replied:
Ah nice!

The-Compiler reviewed (changes requested):
That was quick! :+1:

The spelling check failure is my bad :see_no_evil: - fixed in master, so it should go away if you rebase again.

The-Compiler commented:
Hm, looks like `completions:` only works for the `String` type... here is a quick fix:

```diff
diff --git i/qutebrowser/config/configdata.yml w/qutebrowser/config/configdata.yml
index b9d22a174..f9f14e95c 100644
--- i/qutebrowser/config/configdata.yml
+++ w/qutebrowser/config/configdata.yml
@@ -1222,8 +1222,8 @@ fileselect.single_file.command:
     name: ShellCommand
     placeholder: true
     completions:
-      - ['xterm', '-e', 'ranger --choosefile={}']
-      - ['xterm', '-e', 'vifm', '--choose-file', '{}']
+      - ['["xterm", "-e", "ranger --choosefile={}"]', "Ranger in xterm"]
+      - ['["xterm", "-e", "vifm", "--choose-file", "{}"]', "vifm in xterm"]
   default: ['xterm', '-e', 'ranger --choosefile={}']
   desc: >-
     Command (and arguments) to use for selecting a single file in forms.
@@ -1238,8 +1238,8 @@ fileselect.multiple_files.command:
     name: ShellCommand
     placeholder: true
     completions:
-      - ['xterm', '-e', 'ranger --choosefile={}']
-      - ['xterm', '-e', 'vifm', '--choose-files', '{}']
+      - ['["xterm", "-e", "ranger --choosefiles={}"]', "Ranger in xterm"]
+      - ['["xterm", "-e", "vifm", "--choose-files", "{}"]', "vifm in xterm"]
   default: ['xterm', '-e', 'ranger --choosefiles={}']
   desc: >-
     Command (and arguments) to use for selecting multiple files in forms.
diff --git i/qutebrowser/config/configtypes.py w/qutebrowser/config/configtypes.py
index 6328c3140..fde4a24f0 100644
--- i/qutebrowser/config/configtypes.py
+++ w/qutebrowser/config/configtypes.py
@@ -1522,9 +1522,17 @@ class ShellCommand(List):
     _show_valtype = False
 
     def __init__(self, placeholder: bool = False,
+                 completions: _Completions = None,
                  none_ok: bool = False) -> None:
         super().__init__(valtype=String(), none_ok=none_ok)
         self.placeholder = placeholder
+        self._completions = completions
+
+    def complete(self) -> _Completions:
+        if self._completions is not None:
+            return self._completions
+        else:
+            return super().complete()
 
     def to_py(
             self,

```

I'll look at a proper fix (or open an issue instead probably) somewhen later.

AckslD commented:
@The-Compiler I see! I'll keep an eye out for the issue to fix the `completion:` for other types.

Other than that the last remaining point is to check if there indeed is an issue with the opened temporary file on windows as you mentioned. Did you manage to try this out?

The-Compiler commented:
> Other than that the last remaining point is to check if there indeed is an issue with the opened temporary file on windows as you mentioned. Did you manage to try this out?

Not yet - I'd also like to write an end2end test or two for the file-selection as this is something we can do now. That'd most likely also answer that question as well.

If you want to take a stab at it, `tests/end2end/features/test_editor_bdd.py` and `editor.feature` would be the place to look at. If not, I completely understand, and I'd happily take care of it - though most likely after I finished #6019.

AckslD commented:
@The-Compiler Sounds good :) I can maybe take a look at it, however not before ~end of next week.

The-Compiler commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,37 @@ def netrc_authentication(url, authenticator):
+
+        loop = qtutils.EventLoop()
+        proc.finished.connect(lambda _code, _status: loop.exit())
+        loop.exec_()
```
This should be `.exec()`, see 99017c50a.

AckslD replied:
Changed

AckslD commented:
@The-Compiler I started looking into how one writes the end2end tests, see the WIP files `tests/end2end/features/fileselect.feature` and `tests/end2end/features/test_fileselect_bdd.py`. What I am unsure about is how to trigger the fileselection in the first place and how to check that the correct file is chosen/uploaded, using the fake command used in the test. Any ideas/tips?

The-Compiler commented:
The missing piece is a HTML page with a file input, which then is in `tests/end2end/data/` somewhere. See the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file) - at the bottom, it has a (rather complex) example with the `updateImageDisplay` function which you can simplify (or just use as an inspiration).

That page would then take the selected file and either update some element on the page, or log something via `console.log` in JS. Then the test can either get that element via `:jseval` or just check the log output. 

For some inspiration on how the "big picture" would look, see e.g. an end2end test I added recently in fee09462965e8f19f067a467f85b00cd362ddff6 with `tests/end2end/data/misc/xhr_headers.html` and the "Custom headers via XHR" test in `tests/end2end/features/misc.feature`.

AckslD commented:
Thanks! I will take a look at this :+1:

AckslD commented:
@The-Compiler I couldn't find an existing HTML file which triggers a file choose dialog in `tests/end2end/data/`. I tried to create one with the following HTML code

```HTML
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <title>Testing fileselection</title>
        <script type="text/javascript">
            function onLoad() {
                const input = document.querySelector('#file_upload');
                input.click()
            }
        </script>
    </head>
    <body onload="onLoad()">
        <input type="file" id="file_upload" name="file_upload" accept=".txt" single>
    </body>
</html>
```

However, this gives the warning `File chooser dialog can only be shown with a user activation.`. The same happens if the "click" happens using `jseval` and also for custom fileselectors. So I'm not sure how to trigger the file selector without user activation. Perhaps through the python test code in some way?

The-Compiler commented:
You should be able to click it from the test code using the `:click-element id file_upload` command (if you search for `click-element` in the feature files, you should find various examples). If that doesn't work either, try `:click-element --force-event` instead.

AckslD commented:
@The-Compiler That works :) I think I'm almost there now. Things work when I manually set the command and trigger the dialog on the test page I created. Running the tests seems to also mostly work except I don't get the javascript logging I expect and I'm not completely sure why.

I can see that `qutebrowser.browser.shared.choose_file` is called since there is one test where a single file should be chosen but the external file selector actually provides two. The expected warning is then raised. However, the javascript logging message from line 10 in `tests/end2end/data/fileselect/single_fileselect.html` is not shown for some reason.

Btw, when testing there is still an issue with `:completions` but I comment them out when testing locally for now.

Not actually sure what is going on. I will try to investigate but if you have some direct ideas feel free to let me know :)

The-Compiler commented on tests/end2end/data/fileselect/static_fileselector.sh:
```diff
@@ -0,0 +1,3 @@
+for file in $1; do
+    echo $file >> $2
+done
```
It'd be great to do this in Python instead (perhaps via `shutils.copyfileobj`) so that running the tests on Windows doesn't require a working `bash` install.

There is already a `py_proc` fixture you can use (see e.g. `tests/unit/commands/test_userscripts.py`) where you can pass Python code as a string, and you get a list like `["python3", "-c", "yourcode"]` back which you can pass as process directly.

AckslD replied:
Sounds good! I tried using the `py_proc` fixture as follows:
```python
FILE_SELECTOR_SCRIPT = r"""
import sys
tmp_file = sys.argv[1]
choosenfiles = {}
with open(tmp_file, 'w') as f:
    for choosenfile in choosenfiles:
        f.write(choosenfile)
"""

@bdd.when(bdd.parsers.parse('I set up a fake single file fileselector '
                            'selecting "{choosefile}"'))
def set_up_single_fileselector(quteproc, py_proc, choosefile):
    """Set up fileselect.single_file.command to select the file."""
    cmd, args = py_proc(FILE_SELECTOR_SCRIPT.format(repr([choosefile])))
    fileselect_cmd = json.dumps([cmd, *args, '{}'])
    quteproc.set_setting('fileselect.handler', 'external')
    quteproc.set_setting('fileselect.single_file.command', fileselect_cmd)
```

However, I get the error `qutebrowser.api.cmdutils.CommandError: Invalid value '...' - may not contain unprintable chars!`. Is this because of newlines or single quotes maybe? Sorry if I perhaps misunderstood what you intended :)

The-Compiler replied:
Hmm, weird. Should work on first sight. Perhaps I should rewrite `py_proc` so it uses a temporary `.py` file with more complicated multiline input. Can you push it so I can take a closer look?

AckslD replied:
Pushed, see e241c8c49

AckslD replied:
Btw, I also restructured some things in 1b2fd8b in order to reuse code. However, the functionality is the same and the issue is still there

The-Compiler commented on tests/end2end/data/fileselect/single_fileselect.html:
```diff
@@ -0,0 +1,20 @@
+            function onLoad() {
+                const input = document.querySelector('#file_upload');
+                input.addEventListener('input', function (e) {
+                    if (input.files.length == 1) {
```
Perhaps log something here if the length isn't 1, so that there's some information if that happens rather than a (rather silent) failure? Perhaps that'll also tell you more about why your test isn't working at the moment.

AckslD replied:
Good point, I'll add that!

AckslD replied:
Done

The-Compiler commented on tests/end2end/data/fileselect/single_fileselect.html:
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
```
I wonder if it makes sense for this to be in two separate files? Might be easier to have two file selects in the same HTML file, so that you also can share some of the JS code around it?

AckslD replied:
Yep, I'll try to make a single file :+1:

AckslD replied:
Done

The-Compiler commented on tests/end2end/data/fileselect/multiple_fileselect.html:
```diff
@@ -0,0 +1,20 @@
+            function onLoad() {
+                const input = document.querySelector('#file_upload');
+                input.addEventListener('input', function (e) {
+                    for (i = 0; i < input.files.length; i++) {
```
Perhaps you could use something like

```json
console.log(`Files: ${JSON.stringify(input.files)}`)
```

here, so that you get a single list of all files? That'd then also simplify the Python side of the test a bit, and make it more robust at the same time, as you can make sure there's nothing unexpected in there.

It'd likely also allow fully sharing the code between the two variants, see my comment below.

AckslD replied:
Sounds good. The only other difference is if the input button should have the `multiple` boolean flag or not. But maybe there can be two buttons with different ids, one for single and one for multiple but they both trigger the same JS function.

The-Compiler replied:
Yep, that should work!

AckslD replied:
I've updated this. However seems `input.files` is a more complex structure, a dictionary of objects? I haven't done much JS before but in python I would do something like `[file.name for file in input.files.values]`. What's a good way to do something similar in JS? :)

The-Compiler reviewed (changes requested):
Awesome, thank you! I'm not entirely sure why it's failing, but I added a couple of comments on the tests.

The-Compiler commented:
> Btw, when testing there is still an issue with `:completions` but I comment them out when testing locally for now.

I think you haven't applied my patch from https://github.com/qutebrowser/qutebrowser/pull/5987#issuecomment-754859070 yet, right? 

As for the `py_proc` issue, here's another patch which helps:

```diff
diff --git i/tests/helpers/fixtures.py w/tests/helpers/fixtures.py
index b814a6ea7..7042a39f5 100644
--- i/tests/helpers/fixtures.py
+++ w/tests/helpers/fixtures.py
@@ -503,13 +503,19 @@ def cookiejar_and_cache(stubs, monkeypatch):
 
 
 @pytest.fixture
-def py_proc():
+def py_proc(tmp_path):
     """Get a python executable and args list which executes the given code."""
     if getattr(sys, 'frozen', False):
         pytest.skip("Can't be run when frozen")
 
     def func(code):
-        return (sys.executable, ['-c', textwrap.dedent(code.strip('\n'))])
+        code = textwrap.dedent(code.strip('\n'))
+        if '\n' in code:
+            py_file = tmp_path / 'py_proc.py'
+            py_file.write_text(code)
+            return (sys.executable, [str(py_file)])
+        else:
+            return (sys.executable, ['-c', code])
 
     return func
 
```

With that, I now get:

```
17:56:41.145 WARNING  qt         Unknown module:none:0 Ignoring invalid item in FilePickerController::accepted(QStringList): tests/end2end/data/numbers/1.txt
```

and

```
17:56:34.457 WARNING  qt         Unknown module:none:0 Ignoring invalid item in FilePickerController::accepted(QStringList): tests/end2end/data/numbers/1.txt tests/end2end/data/numbers/2.txt
```

I think for this to work properly, those should be absolute rather than relative paths?

AckslD commented:
Great! :) I've applied the patch and also the earlier one for completions.

I think you're right that they need to be absolute paths. I guess the path should be expanded in the python test file and leave the relative path in the .feature file?

AckslD commented:
Hmm, I don't see those warnings you mentioned. They do come up if I manually test the page (with non-existant files) but not when I run the tests using tox. How did you produced them?

The-Compiler commented:
> I think you're right that they need to be absolute paths. I guess the path should be expanded in the python test file and leave the relative path in the .feature file?

Agreed! While you're at it, perhaps it'd also make sense to pass those via `sys.argv` (rather than formatting them into the script) to make things a bit more straightforward?

> Edit: I read you comment to quick and missed the author part.

No worries!

> Hmm, I don't see those warnings you mentioned. They do come up if I manually test the page (with non-existant files) but not when I run the tests using tox. How did you produced them?

I did run `tox -e py39-pyqt515 -- tests/end2end/features/test_fileselect_bdd.py`

Also, the linter failures are my fault - should go away if you do a merge from master.

AckslD commented:
I fixed that absolute paths and passing the selected files through `sys.argv` instead.

Something is still not working correctly for some reason and I was not seeing the expected warnings before using absolute paths.

The-Compiler reviewed (changes requested):
I commented on two bugs in the test code I've noticed.

With those changes I now get `Files: {"0":{}}` from JS. I have no idea why... One possible explanation is that the Python part is actually working properly now, but [FileList](https://developer.mozilla.org/en-US/docs/Web/API/FileList) and [File](https://developer.mozilla.org/en-US/docs/Web/API/File) don't serialize to JSON properly in JS. You might need to still have a loop there which first creates an array of strings (the filenames only), and then serialize that instead.

The-Compiler commented on tests/end2end/features/test_fileselect_bdd.py:
```diff
@@ -0,0 +1,63 @@
+    set_up_fileselector(
+        quteproc=quteproc,
+        py_proc=py_proc,
+        choosefiles=choosefiles,
```
Huh, somehow this comment got lost?!

```suggestion
        choosefiles=choosefiles.split(' '),
```

otherwise you're passing the raw string, and the `*` expansion causes this to happen:

`:set -t "fileselect.multiple_files.command" "[\"/__w/qutebrowser/qutebrowser/.tox/py/bin/python\", \"/tmp/pytest-of-user/pytest-0/test_select_two_files_with_mul0/py_proc.py\", \"{}\", \"t\", \"e\", \"s\", \"t\", \"s\", \"/\", \"e\", \"n\", \"d\", \"2\", \"e\", \"n\", \"d\", \"/\", \"d\", \"a\", \"t\", \"a\", \"/\", \"n\", \"u\", \"m\", \"b\", \"e\", \"r\", \"s\", \"/\", \"1\", \".\", \"t\", \"x\", \"t\", \" \", \"t\", \"e\", \"s\", \"t\", \"s\", \"/\", \"e\", \"n\", \"d\", \"2\", \"e\", \"n\", \"d\", \"/\", \"d\", \"a\", \"t\", \"a\", \"/\", \"n\", \"u\", \"m\", \"b\", \"e\", \"r\", \"s\", \"/\", \"2\", \".\", \"t\", \"x\", \"t\"]"`

The-Compiler commented on tests/end2end/data/fileselect/fileselect.html:
```diff
@@ -0,0 +1,21 @@
+                for (let id of ['#single_file', '#multiple_files']) {
+                    const input = document.querySelector(id);
+                    input.addEventListener('input', function (e) {
+                        console.log(`Files: ${JSON.stringify(input.files)}`);
```
```suggestion
                        console.log(`Files: ${JSON.stringify(Array.from(input.files).map(f => f.name))}`);
```

there we go - still nicer than a for-loop in my opinion :slightly_smiling_face:

The-Compiler replied:
Wait, actually no need for JSON at this point really:

```suggestion
                        console.log(`Files: ${Array.from(input.files).map(f => f.name).join(", ")}`);
```

The-Compiler commented on tests/end2end/features/fileselect.feature:
```diff
@@ -0,0 +1,36 @@
+        And I open data/fileselect/fileselect.html
+        And I run :click-element id single_file
+        Then the javascript message "File selected: 1.txt" should be logged
+        Then the javascript message "File selected: 2.txt" should not be logged
```
This should now be `Files: 1.txt, 2.txt` (same for the other places in here as well).

AckslD replied:
True! Although in this scenario it should be only `Files: 1.txt` right? Since we return only the first entry if the `single_file.command` wrongly returns multiple files. I also fixed another small issue with the test file in this case

The-Compiler replied:
Err, yes, I commented on the wrong line.

AckslD replied:
I think things might be working now, let's see if the pipeline pass. Btw, does it declare failure in the case of a warning being raised? Since the above scenario will raise a warning. Got me thinking, could that warning be asserted perhaps?

The-Compiler replied:
> Btw, does it declare failure in the case of a warning being raised? Since the above scenario will raise a warning. Got me thinking, could that warning be asserted perhaps?

I guess the failing tests answer that question :wink:

You can add `And the warning "..." should be shown` to the "Then" block to mark it as expected (and also assert it).

The-Compiler commented on qutebrowser/browser/shared.py:
```diff
@@ -339,3 +343,37 @@ def netrc_authentication(url, authenticator):
+        If multiple is False, the return value will have at most 1 item.
+    """
+    encoding = sys.getfilesystemencoding()
+    with tempfile.NamedTemporaryFile(mode='w+', encoding=encoding) as tmpfile:
```
Well, looks like this indeed doesn't work out on Windows:

```
  18:59:41.412 ERROR    procs      guiprocess:_on_finished:125 Process stderr:
  Traceback (most recent call last):
  
    File "C:\Users\runneradmin\AppData\Local\Temp\pytest-of-runneradmin\pytest-0\test_select_two_files_with_sin0\py_proc.py", line 4, in <module>
  
      with open(tmp_file, 'w') as f:
  
  PermissionError: [Errno 13] Permission denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpsguwnqtk'
```

I should've known... see #1789 and #1927 :laughing:

So I'd suggest something like here instead:

https://github.com/qutebrowser/qutebrowser/blob/a17a2ad1738c2483bed6c24a4a15fcb8dfaf49f8/qutebrowser/commands/userscripts.py#L339-L345

(perhaps also add `prefix='qutebrowser-fileselect-'` - I wonder why we don't have that for userscripts actually)

But you'll also need to take care that the file is actually deleted, even if an exception happens or we return early. The nicest way is probably a custom context manager:

```python3
import contextlib

@contextlib.contextmanager
def cleanup_file(filename):
    try:
        yield filename
    finally:
        try:
            os.remove(filename)
        except OSError as e:
            message.error(f"Failed to delete tempfile {filename} ({e})!")
```

(untested)

Which you can then use like:

```python3
with cleanup_file(filename):
    # ...
```

Would be nice to have a unit-test or two for the context manager as well, but I can take care of that if you prefer.

AckslD replied:
Sounds good :+1:! Where would you prefer the context manager (+tests) to live?

The-Compiler replied:
I'd say probably in `utils/utils.py`.

AckslD replied:
A question regarding windows btw, in the `.feature`-file we specify the path to the file to select using `/` but this is I guess also an issue in windows? One way would be to in the python test script do something like
```python
os.path.join(*file.split('/'))
```
but maybe there is a nicer way to handle that.

The-Compiler replied:
Nope - they actually get served via a HTTP server started on localhost (`tests/end2end/fixtures/webserver_sub.py`), so something like "I open data/fileselect.html" actually turns into a URL like http://localhost:1234/data/fileselect.html

AckslD replied:
I added the `cleanup_file` context and three tests. Two questions though:

1. If the file does not exist, should the context log an error? Alternatively we can check if it actually exists before trying to delete it.
2. Can the error in case of failure be done with `log.misc.error`? I tried to use `message.error` but then got a circular import.

The-Compiler replied:
1. I think that'd be covered by the `except OSError:` already, as `FileNotFoundError` is a subclass of it
2. Oh, that's unfortunate. Yeah, feel free to use that instead, I guess it doesn't happen often and it's not something very relevant for the user anyways.

AckslD replied:
Re 1. It is indeed covered, but maybe an error should not be logged in that case? It might be confusing if it says that a temporary file could not be removed when in fact is does not exist. What do you think?

The-Compiler commented on qutebrowser/config/configtypes.py:
```diff
@@ -1522,9 +1522,17 @@ class ShellCommand(List):
+
+    def complete(self) -> _Completions:
+        if self._completions is not None:
+            return self._completions
```
coverage complains because this line isn't tested - as this will be refactored soon-ish anyways, feel free to add a `# pragma: no cover` for now.

AckslD replied:
Also the line when the `default` file selector is used is not covered but not sure how to test that case
https://github.com/qutebrowser/qutebrowser/blob/965062c0b6800cfa3809f1171e7f8f7541626569/qutebrowser/browser/webengine/webview.py#L251

The-Compiler replied:
Yep - there's no good way to test that automatically (I think I tried a while back, and ran into issues with not being able to close the file selectors, since they're not Qt widgets but native Windows/GTK/... things).

However, qutebrowser enforces 100% coverage for files which are typically easy to test (see `PERFECT_FILES` in `scripts/dev/check_coverage.py`). `configtypes.py` is one of those, but `webengine/webview.py` isn't, so this is needed here but not there.

The-Compiler commented on tests/end2end/data/fileselect/fileselect.html:
```diff
@@ -0,0 +1,21 @@
+            function onLoad() {
+                for (let id of ['#single_file', '#multiple_files']) {
+                    const input = document.querySelector(id);
+                    input.addEventListener('input', function (e) {
```
Looks like QtWebKit doesn't support [`input` events](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event) on `file` inputs - it does however seem to support [`change` events](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event), which for our purposes should be equivalent.

Thus, this should fix the QtWebKit failure (and still work on QtWebEngine):

```suggestion
                    input.addEventListener('change', function (e) {
```

AckslD replied:
Changed

The-Compiler commented on tests/end2end/features/fileselect.feature:
```diff
@@ -0,0 +1,35 @@
+# vim: ft=cucumber fileencoding=utf-8 sts=4 sw=4 et:
```
For every `.feature` file, a new instance of qutebrowser is spanwed (which then runs all the tests in that file). This is to have some isolation between tests, but still have acceptable performance. Because we aren't going to have many tests here, I'd prefer having them in `editor.feature` instead of having them in their own file. It's not a perfect match of course, but I'd say they fit in best there.

AckslD replied:
Merged with `editor.feature`

The-Compiler commented on tests/end2end/features/fileselect.feature:
```diff
@@ -0,0 +1,35 @@
+Feature: Selecting files
+
+    Background:
+        Given I have a fresh instance
```
This slows down things a lot because it starts a new instance for every test. I just noticed this is in `editor.feature` as well though (via 6519f500a9f779ce868d1dd9556902e607cde6ab)...

It's not actually needed, and removing it as much as possible (62cf48fc779a22617957bd18272a55a04a8ffc2d) now makes the editor tests run in 8s instead of 30s on my machine...

Probably irrelevant for you with the comment above, though.

AckslD replied:
Yep, since merged with `editor.feature`

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -238,3 +239,16 @@ def acceptNavigationRequest(self,
+        """Override chooseFiles to (optionally) invoke custom file uploader."""
+        if config.val.fileselect.handler == "default":
+            return super().chooseFiles(mode, old_files, accepted_mimetypes)
+
```
Perhaps add `assert handler == "external", handler` here, so this fails rather than doing the wrong thing if we ever add a new value there and forget to handle it here.

(At that point it probably makes sense again to do `handler = config.val.fileselect.handler` - sorry for that!

AckslD replied:
Done

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -192,6 +192,17 @@ def _handle_errorpage(self, info, errpage):
+        """Override chooseFile to (optionally) invoke custom file uploader."""
+        if config.val.fileselect.handler == "default":
+            return super().chooseFile(parent_frame, suggested_file)
+
```
Same as above regarding the assertion.

AckslD replied:
Done

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -192,6 +192,17 @@ def _handle_errorpage(self, info, errpage):
+        if config.val.fileselect.handler == "default":
+            return super().chooseFile(parent_frame, suggested_file)
+
+        selected_file = shared.choose_file(multiple=False)
```
nitpick: I see why you're calling this `selected_file`, but it sounds like this would be a single file rather than a list containing one file. Maybe just call it `selected`?

AckslD replied:
Changed to `selected`

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -205,13 +216,16 @@ def _handle_multiple_files(self, info, files):
+            suggested_file = ""
+            if info.suggestedFileNames:
+                suggested_file = info.suggestedFileNames[0]
 
```
Same as above regarding the assertion.

AckslD replied:
Done

The-Compiler commented on tests/end2end/data/fileselect/fileselect.html:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
```
Since we now have a single `fileselect.html`, I suppose it doesn't make much sense to have it in a subfolder.

AckslD replied:
Moved file and one level up

The-Compiler commented on tests/end2end/features/test_fileselect_bdd.py:
```diff
@@ -0,0 +1,62 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016-2020 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
```
```suggestion
# Copyright 2020 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
```

Probably irrelevant with the comment above though (and I should really update those for 2021...)

AckslD replied:
Yep, since merged with editor tests

The-Compiler commented on tests/end2end/features/test_fileselect_bdd.py:
```diff
@@ -0,0 +1,62 @@
+    set_up_fileselector(
+        quteproc=quteproc,
+        py_proc=py_proc,
+        choosefiles=choosefiles,
+        setting_cmd='fileselect.single_file.command'
```
I'm a bit confused about those names (both here and below). I'd prefer something like:

- `selected_files` or perhaps even just `files` (rather than `choosefiles`)
- `setting` or `setting_name` (it's not a command)

AckslD replied:
Fixed

The-Compiler commented on tests/end2end/features/test_fileselect_bdd.py:
```diff
@@ -0,0 +1,62 @@
+bdd.scenarios('fileselect.feature')
+
+
+@bdd.when(bdd.parsers.parse('I set up a fake single file fileselector '
```
With

```suggestion
@bdd.when(bdd.parsers.parse('I set up a fake {kind} file fileselector '
```

You could unify those three functions into one, which then uses `kind == "multiple"` or so to get the setting. Not sure how much of an improvement it'd be though, but I guess especially if this gets moved to `test_editor_bdd.py` it would make things a bit clearer.

AckslD replied:
Done

The-Compiler commented on tests/end2end/features/test_fileselect_bdd.py:
```diff
@@ -0,0 +1,62 @@
+        import sys
+        tmp_file = sys.argv[1]
+        with open(tmp_file, 'w') as f:
+            for choosenfile in sys.argv[2:]:
+                f.write(os.path.abspath(choosenfile) + "\n")
```
"chosen" is the correct English spelling, but I'd probably just call it "selected_file" for consistency - your call :wink:

AckslD replied:
Fixed

The-Compiler reviewed (changes requested):
Looks like we're getting closer to having this merged - thanks for your patience!

Most of the comments I've added are relatively minor, but unfortunately, like suspected, it doesn't work on Windows yet. I added a comment with a suggestion on how to handle that as well.

AckslD commented:
Yes :) No problem! I'm happy to do this since I'm learning a lot. So thanks for taking your time explaining and helping out!

I'll go through the comments and try to fix the issue on windows :+1:

The-Compiler commented on tests/unit/utils/test_utils.py:
```diff
@@ -878,3 +878,31 @@ def test_parse_duration_hypothesis(duration):
+
+class TestCleanupFileContext:
+
+    def test_no_file(self, tmpdir):
```
Please use `tmp_path` instead, as it's the more modern replacement using the `pathlib.Path` API instead of pytests own one - see https://docs.pytest.org/en/stable/tmpdir.html and #176.

AckslD replied:
Done

The-Compiler commented on tests/unit/utils/test_utils.py:
```diff
@@ -878,3 +878,31 @@ def test_parse_duration_hypothesis(duration):
+        tmpfile = tmpdir / 'tmp.txt'
+        with utils.cleanup_file(tmpfile):
+            pass
+        assert not os.path.exists(tmpfile)
```
```suggestion
        assert not tmpfile.exists()
```

AckslD replied:
Changed, although still keeping `tmpfile = tmp_path / `tmp.txt` or would you prefer to use `tmp_path` directly?

The-Compiler commented on tests/unit/utils/test_utils.py:
```diff
@@ -878,3 +878,31 @@ def test_parse_duration_hypothesis(duration):
+        tmpfile = tmpdir / 'tmp.txt'
+        with tmpfile.open('w'):
+            pass
+        try:
```
Using `with pytest.raises(RuntimeError):` here instead would make sure that the exception is also correctly bubbled through - see https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions

AckslD replied:
I had that at first but my linter then complains that the last line is unreachable. But I didn't actually try if the linter used in the CI also thinks that.

The-Compiler replied:
Did you have the last line outside of the `with`-block? If it was inside, that would've been correct - the code after the exception isn't run.

AckslD replied:
Yes, outside the `with`-block. Seems to maybe be a bug with my linter indeed (`coc-pyright`). Anyway, changed to use `pytest.raises`

AckslD commented:
All seems to work except the coverage failed since the case when the cleanup of the file fails was not tested. I added a test when the path is actually a directory. For now I just added a `caplog.at_level(logging.ERROR, 'misc')` context but is there a better way to assert that the error logged is actually the expected one?

The-Compiler commented on qutebrowser/browser/webengine/webview.py:
```diff
@@ -238,3 +239,18 @@ def acceptNavigationRequest(self,
+        if handler == "default":
+            return super().chooseFiles(mode, old_files, accepted_mimetypes)
+
+        assert handler == "external"
```
```suggestion
        assert handler == "external", handler
```

So that the error printed is "AssertionError: someothervalue" rather than just "AssertionError" - i.e. Python prints the thing after the comma if the assertion fails, which helps with debugging issues:

```pycon
>>> handler = "notext"
>>> assert handler == "external"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
>>> assert handler == "external", handler
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: notext
```

(With pytest, we get nice informations for AssertionErrors without that, but that's a [pytest-specific feature](https://docs.pytest.org/en/stable/assert.html))

AckslD replied:
Done

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -192,6 +192,19 @@ def _handle_errorpage(self, info, errpage):
+        if handler == "default":
+            return super().chooseFile(parent_frame, suggested_file)
+
+        assert handler == "external"
```
```suggestion
        assert handler == "external", handler
```

AckslD replied:
Done

The-Compiler commented on qutebrowser/browser/webkit/webpage.py:
```diff
@@ -205,13 +218,18 @@ def _handle_multiple_files(self, info, files):
+                None, None, suggested_file)  # type: ignore[arg-type]
+            return True
 
+        assert handler == "external"
```
```suggestion
        assert handler == "external", handler
```

AckslD replied:
Done

The-Compiler commented on qutebrowser/utils/utils.py:
```diff
@@ -838,3 +838,20 @@ def mimetype_extension(mimetype: str) -> Optional[str]:
+        yield
+    finally:
+        try:
+            if os.path.exists(filepath):
```
I still don't think this is needed because we handle `OSError` anyways:

```pycon
>>> try:
...     os.remove('/doesnotexist')
... except OSError as e:
...     print(f"Failed to delete file: {e}")
... 
Failed to delete file: [Errno 2] No such file or directory: '/doesnotexist'
```

AckslD replied:
Removed if-statement

The-Compiler commented on tests/unit/utils/test_utils.py:
```diff
@@ -878,3 +878,36 @@ def test_parse_duration_hypothesis(duration):
+        # removal of file fails since it's a directory
+        with caplog.at_level(logging.ERROR, 'misc'):
+            with utils.cleanup_file(tmp_path):
+                pass
```
As for your question regarding checking the log message: See the [caplog documentation](https://docs.pytest.org/en/stable/logging.html#caplog-fixture) (and perhaps also [API reference](https://docs.pytest.org/en/stable/reference.html#pytest.LogCaptureFixture).

You'll need to be careful because the exact error message will be different on Linux/Windows. Perhaps something like this would suffice:

```python
assert len(caplog.messages) == 1
assert caplog.messages[0].startswith("Failed to delete tempfile")
```

AckslD replied:
Added this and also for the test when there is no such file

The-Compiler reviewed (changes requested):
Yay, looks like tests are green! :tada: 

Only a couple of small issues remaining now.

The-Compiler commented:
Looks great now! Merged with a couple of minor changes:

- d47f87132 Fix file select warning
- 1a6b69eb2 Update docs/changelog

Thanks @user202729 for the original PR, and thanks @AckslD for picking it up and taking care of everything! :+1:

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -1201,6 +1201,55 @@ editor.encoding:
+    placeholder: true
+    completions:
+      - ['["xterm", "-e", "ranger --choosefile={}"]', "Ranger in xterm"]
+      - ['["xterm", "-e", "vifm", "--choose-file", "{}"]', "vifm in xterm"]
```
FYI, looks like `vifm --chose-file` doesn't actually exist... Changed to `--choose-files` in 8faa60592.

AckslD replied:
Indeed seems like `vifm --choose-file` silently fallbacks to `vifm --choose-files`.
